linera-views/src/storage_metric.rs

```diff
@@ -0,0 +1,287 @@
+    pub n_reads: usize,
+    /// The number of missed cases in read_key and read_multi_key
+    pub n_miss_reads: usize,
+    /// The number of Put in the batches
+    pub n_puts: usize,
+    /// The total number of Delete in the batches
+    pub n_deletes: usize,
+    /// The total number of DeletePrefix in the batches
+    pub n_delete_prefix: usize,
+    /// The total data that went into the Batches
+    pub size_writes: usize,
+    /// The total data being read from
+    pub size_reads: usize,
+}
+
+/// The `MetricKeyValueClient` encapsulates a client and creates a new one that
+/// measures the operations being done
+#[derive(Clone)]
+pub struct MetricKeyValueClient<K> {
+    /// The inner client that is called by the metric one
+    pub client: K,
+    /// The data contained in the running of this container
+    pub metric_stat: Arc<RwLock<MetricStat>>,
+}
+
+#[async_trait]
+impl<K> KeyValueStoreClient for MetricKeyValueClient<K>
+where
+    K: KeyValueStoreClient + Send + Sync,
+{
+    const MAX_VALUE_SIZE: usize = K::MAX_VALUE_SIZE;
+    type Error = K::Error;
+    type Keys = K::Keys;
+    type KeyValues = K::KeyValues;
+
+    fn max_stream_queries(&self) -> usize {
+        self.client.max_stream_queries()
+    }
+
+    async fn read_key_bytes(&self, key: &[u8]) -> Result<Option<Vec<u8>>, Self::Error> {
+        let read = self.client.read_key_bytes(key).await?;
+        let mut metric_stat = self.metric_stat.write().await;
+        metric_stat.n_reads += 1;
+        match &read {
+            None => {
+                metric_stat.n_miss_reads += 1;
+            }
+            Some(value) => {
+                metric_stat.size_reads += value.len();
+            }
+        }
+        Ok(read)
+    }
+
+    async fn read_multi_key_bytes(
+        &self,
+        keys: Vec<Vec<u8>>,
+    ) -> Result<Vec<Option<Vec<u8>>>, Self::Error> {
+        let n_keys = keys.len();
+        let multi_read = self.client.read_multi_key_bytes(keys).await?;
+        let mut metric_stat = self.metric_stat.write().await;
+        metric_stat.n_reads += n_keys;
+        for read in &multi_read {
+            match read {
+                None => {
+                    metric_stat.n_miss_reads += 1;
+                }
+                Some(value) => {
+                    metric_stat.size_reads += value.len();
+                }
+            }
+        }
+        Ok(multi_read)
+    }
+
+    async fn find_keys_by_prefix(&self, key_prefix: &[u8]) -> Result<Self::Keys, Self::Error> {
+        let mut metric_stat = self.metric_stat.write().await;
+        let keys = self.client.find_keys_by_prefix(key_prefix).await?;
+        for key in keys.iterator() {
+            metric_stat.size_reads += key?.len();
+        }
+        Ok(keys)
+    }
+
+    async fn find_key_values_by_prefix(
+        &self,
+        key_prefix: &[u8],
+    ) -> Result<Self::KeyValues, Self::Error> {
+        let mut metric_stat = self.metric_stat.write().await;
```

Comment:
This is taking the lock and keeping it during I/O operations.

Reply:
Ok, addressed with atomics as suggested.